src/lib.rs

```diff
@@ -0,0 +1,105 @@
+extern crate libc;
+extern crate rand;
+
+use libc::c_void;
+use rand::Rng;
+use std::cmp;
+use std::env;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::{self, AtomicUsize};
+
+const CANARY_SIZE: usize = mem::size_of::<usize>();
+
+static MEM_INIT: AtomicUsize = atomic::ATOMIC_USIZE_INIT;
+static CONF_ALLOC_EXTRA_MEM: AtomicUsize = atomic::ATOMIC_USIZE_INIT;
+
+pub extern fn libdiffuzz_init_config() {
+    if env::var_os("LIBDIFFUZZ_NONDETERMINISTIC").is_some() {
+        MEM_INIT.store(
+            rand::thread_rng().gen::<u8>().into(),
+            atomic::Ordering::Relaxed,
+        );
+    }
+    let alloc_extra_mem = env::var("LIBDIFFUZZ_ALLOCATE_EXTRA_MEMORY")
+        .ok()
+        .and_then(|x| x.parse().ok())
+        .unwrap_or(0);
+    CONF_ALLOC_EXTRA_MEM.store(alloc_extra_mem, atomic::Ordering::Relaxed);
+}
+
+#[link_section = ".ctors"]
+pub static CONSTRUCTOR: extern fn() = libdiffuzz_init_config;
+
+/// Gets then increments MEM_INIT
+fn get_mem_init() -> u8 {
+    MEM_INIT.fetch_add(1, atomic::Ordering::Relaxed) as u8
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn malloc(len: usize) -> *mut c_void {
+    let alloc_extra_mem = CONF_ALLOC_EXTRA_MEM.load(atomic::Ordering::Relaxed);
+    let full_len = len + CANARY_SIZE + alloc_extra_mem;
```

Comment:
I believe this can overflow and return a buffer that's smaller than the requested amount. There should probably be an overflow check here, returning NULL in case of overflow.

On a somewhat related note, I don't think unwinding is appropriate for a dynamic library injected via LD_PRELOAD, so please add the following to Cargo.toml:
```
[profile.release]
panic = "abort"
```

Reply:
Good points. I'll fix those (and other instances of possible overflow).